gitindex/index.go

```diff
@@ -532,6 +539,229 @@ func newIgnoreMatcher(tree *object.Tree) (*ignore.Matcher, error) {
 	return ignore.ParseIgnoreFile(strings.NewReader(content))
 }
 
+type gitIndexConfig struct {
```

Comment:
nit: this config adds a lot of visual noise at the call sites of indexGitRepo. Maybe using the funcs directly is easier to read?

```
type prepareNormalBuildFunc ...
type prepareDeltaBuildFunc ...

// call site
indexGitRepo(opts, prepareNormalBuild, prepareDeltaBuild)
```

@keegancsmith WDYT?

Reply:
> Yeah. Alternatively can't we setup git repos which emulate the behaviours we want to test for instead of mocking? That seems more robust. But unsure of the difficulty of that.

I'm not sure what exactly you mean here? The mock strategy allows to us to expect externally whether or not a delta build and/or a normal build was use to produce the shards. Otherwise, there isn't a way for us to externally get this information. 

The mocks don't affect / modify the contents of the git repository we're indexing. I'm not sure how creating special git repos would help with the above problem.

Reply:
> inspect externally whether or not a delta build and/or a normal build was use to produce the shards.

In general I find spies the assert something was called to be brittle tests, since they are so dependent on code structure vs actual outcomes. Could you not just assert the presence of a delta shard to determine this?

I realise though that depending on return values from the prepareDeltaBuild we adjust what we do and you want to test that. So you can't really make a self contained pure function which decides which build to do based on inputs. Lets spend some time discussing this today :)